packages/thrift-client/src/main/connections/HttpConnection.ts

```diff
@@ -217,56 +214,47 @@ export class HttpConnection extends Core.ThriftConnection<RequestOptions> {
                 : this.url
 
         // Merge user options with required options
-        const requestOptions: RequestOptions & UrlOptions = Core.overlayObjects(
-            this.requestOptions,
-            options,
+        const requestOptions: RequestOptions = mergeOptions(
+            mergeOptions(this.requestOptions, options),
             {
                 method: 'POST',
                 body: dataToWrite,
-                encoding: null, // Needs to be explicitly set to null to get Buffer in response body
-                url: requestUrl,
                 headers: {
-                    'Content-Length': dataToWrite.length,
                     'Content-Type': 'application/octet-stream',
                 },
             },
         )
 
         return new Promise((resolve, reject) => {
-            this.requestImpl(
-                requestOptions,
-                (err: any, response: RequestResponse, body: Buffer) => {
+            this.requestImpl(requestUrl, requestOptions)
+                .then(response => {
                     if (
                         shouldRetry(response, retry, this.withEndpointPerMethod)
                     ) {
                         resolve(
                             this.write(dataToWrite, methodName, options, true),
                         )
+                    } else if (isErrorResponse(response)) {
+                        reject(response)
```

Comment:
What are your thoughts on making a new Error type that inherits `Error` and has a response as an attribute? Just to keep the interface an error-like object.

Reply:
We absolutely should throw `Error` or something that inherits `Error` so there's a stacktrace. Also it's just a polite thing to do. I filed https://github.com/creditkarma/thrift-server/issues/166 for this (and we have an issue in our tracker at Doma, too).

But I think it's best done as a separate PR. From the point of view of users of this library that's a significant change. Putting it in its own PR increases its visibility and makes it easier for people to review it.

Reply:
Yep, my goal with this PR was to only handle the switch from `request` to `cross-fetch` without introducing a different breaking change, so I'll save that for a different day. Thanks for catching it and bringing it up!